include/dlaf/communication/async.h

```diff
@@ -0,0 +1,108 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+/// @file
+
+#include <utility>
+#include <vector>
+
+#include <mpi.h>
+#include <hpx/async.hpp>
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/include/threads.hpp>
+#include <hpx/lcos/future.hpp>
+#include <hpx/util/yield_while.hpp>
+
+#ifdef DLAF_WITH_MPI_FUTURES
+#include <hpx/mpi/mpi_executor.hpp>
+#include <hpx/mpi/mpi_future.hpp>
+#endif
+
+namespace dlaf {
+namespace comm {
+
+/// The function only accepts non-blocking MPI routines. The MPI library needs
+/// to support MPI_THREAD_MULTIPLE.
+///
+/// Example usage:
+///
+/// dlaf::comm::async(ex, comm, MPI_Irecv, recv_ptr, num_recv_elems, MPI_DOUBLE, src_rank, tag);
+///
+/// compared to the usual MPI non-blocking call:
+///
+/// MPI_Irecv(recv_ptr, num_recv_elems, MPI_DOUBLE, src_rank, tag, comm, &request);
+///
+/// there are two differences
+///
+/// 1) MPI_Comm (comm) goes after the executor in async()
+/// 2) MPI_Requst (request) is omitted as it is handled internally in async()
+///
+/// Note that if starting the communication depends on a previous task, it is often a good idea to call
+/// the `get()` method on a returned future to avoid weird constructs like futute<future<>>.
+///
+/// Example:
+///
+/// auto comm_fut = prev_fut.then([=] {
+///   return dlaf::comm::async(...).get();
+/// });
+///
+/// or don't return the future from `dlaf::comm::async()` and just do:
+///
+/// auto comm_fut = prev_fut.then([=] {
+///   dlaf::comm::async(...);
+/// });
+///
+template <class F, class... Ts>
+hpx::future<void> async(hpx::threads::executors::pool_executor& ex, MPI_Comm comm, F&& f,
+                        Ts&&... ts) noexcept {
+  return hpx::async(ex, [=] {
+    MPI_Request req;
+    f(ts..., comm, &req);
+
+    // Yield until non-blocking communication completes.
+    hpx::util::yield_while([&req] {
+      int flag;
+      MPI_Test(&req, &flag, MPI_STATUS_IGNORE);
+      return flag == 0;
+    });
+  });
+
+/// This section is in preparation for MPI_Futures in HPX 1.5. The mechanism for handling MPI
+/// communication is similar but there are notable differences.
+///
+/// Example usage:
+///
+/// ```
+/// // registers a `poll()` function on the thread pool with `pool_name`. This has to be in scope of all
+/// // uses of hpx::mpi.
+/// hpx::mpi::experimental::enable_user_polling enable_polling(pool_name);
+/// ...
+/// hpx::mpi::experimental::executor ex(mpi_comm);
+/// hpx::async(ex, MPI_Irecv, MPI_Irecv, recv_ptr, num_recv_elems, MPI_DOUBLE, src_rank, tag)
```

Comment:
`MPI_Irecv` is called twice: do we need to add `mpi_comm` as an argument of `async` or this is not necessary anymore since it will be forward by the executor?

Reply:
Yes, that was my mistake, I will delete one of the `MPI_Irecv`.